### DISCUSSION TITLE
v1.1.0 release candidate

### DISCUSSION THREAD

**Discussion by FiloSottile** [FiloSottile/age#424] [General]
<p align="center"><img alt="The age logo, an wireframe of St. Peters dome in Rome, with the text: age, file encryption" width="600" src="https://user-images.githubusercontent.com/1225294/132245842-fda4da6a-1cea-4738-a3da-2dc860861c98.png"></p>

age is a simple, modern and secure file encryption tool, format, and Go library. It features small explicit keys, no config options, and UNIX-style composability.

**v1.1.0-rc.1** is the first release candidate of v1.1.0. Users are encourage to test the new release and especially the new features listed below. Issue or UX reports in advance of the final release are greatly appreciated.

📃 In case you missed it: a new, more polished version of [the age format specification](https://age-encryption.org/v1) has been published.

## Plugin support

The age CLI now supports plugins, such as [age-plugin-yubikey](https://github.com/str4d/age-plugin-yubikey). To test it on macOS with Homebrew:

```
$ brew install --HEAD age
$ brew install age-plugin-yubikey
$ age-plugin-yubikey # interactive setup
$ age -r age1yubikey1qwt50d05nh5vutpdzmlg5wn80xq5negm4uj9ghv0snvdd3yysf5yw3rhl3t
$ age -d -i age-yubikey-identity-388178f3.txt
```

Plugins must be loaded explicitly by using their respective recipient or identity, and are not tied to a specific header stanza type. This means plugins can be used not only to support new recipient types such as PIV tokens (i.e. YubiKeys) or cloud KMS solutions, but also to produce passphrase-encrypted files that can be decrypted without plugins, to store age native private keys on secure elements, or even for agent functionality or to proxy decryption operations to remote machines.

Plugins operate over a simple textual stdin/stdout protocol (https://github.com/C2SP/C2SP/pull/5). Developers are encouraged to [reach out](https://github.com/FiloSottile/age/discussions) with plugin ideas and announcements. Read more in [the relevant man page section](https://htmlpreview.github.io/?https://github.com/FiloSottile/age/blob/main/doc/age.1.html#Plugins).

## CLI breaking changes

**If `-i` is used, passphrase-encrypted files are now rejected.** Previously, a passphrase-encrypted file was auto-detected and the identity file was ignored. This could lead to unexpected behavior, such as a script blocking for user interaction, based on potentially untrusted input files. Now, `age -d` must be invoked without `-i` arguments to decrypt passphrase-encrypted files. A helpful error is printed otherwise. This should not break any automated system as passphrase decryption was always interactive.

**Empty final chunks are now rejected.** If a payload was a multiple of 64KiB long, there were two valid encryptions for it: with a "full" last chunk encrypting 64KiB, or with an additional "empty" chunk encrypting 0 bytes. age, rage, and all other known implementations only ever produced the former. (Note that [age will forever decrypt files it generated](https://htmlpreview.github.io/?https://github.com/FiloSottile/age/blob/main/doc/age.1.html#BACKWARDS-COMPATIBILITY).) The latter is now rejected. The specification is being updated (https://github.com/C2SP/C2SP/pull/13) and test cases will be provided.

## Minor changes

If `/dev/tty` is present but can't be opened, age will now fallback to trying to treat stdin as a terminal as if `/dev/tty` wasn't present.

Windows binary releases are now signed.

Documentation and error messages were improved.

## Learn more

Learn more by reading the [README](https://github.com/FiloSottile/age/blob/main/README.md), the [age(1) man page](https://filippo.io/age/age.1), the [Go API reference](https://pkg.go.dev/filippo.io/age), the [format specification](https://age-encryption.org/v1), or the [full changelog](https://github.com/FiloSottile/age/compare/v1.0.0...v1.1.0-rc.1).

Watch the repository or [follow @FiloSottile on Twitter](https://twitter.com/FiloSottile) to be notified of new releases.

<hr /><em>This discussion was created from the release <a href='https://github.com/FiloSottile/age/releases/tag/v1.1.0-rc.1'>v1.1.0 release candidate</a>.</em>

**Comment by catleeball** [comment]
I'm pretty excited about plugin support! Thanks for your hard work on Age. :)

**Comment by Threnklyn** [comment]
I'm using age as a lib and right now we can use `x25519.go` and `agessh.go` to parse an identity string. but there is no way to parse a plugin identity because all functions regarding this are inside of internal

**Reply by FiloSottile** [reply]
Yeah, we'll expose the plugin API in a following release. There are a couple things I want to clean up before exposing it.

**Reply by Threnklyn** [reply]
great, thanks for your work

**Reply by bexelbie** [reply]
Thank you for the great work.  When do you expect to move this RC to release?

**Comment by dhcgn** [comment]
Plugin cannot be placed in same folder as the age.exe executable (windows), is this intended?

age: error: failed to wrap key for recipient #0: yubikey plugin: couldn't start plugin: age-plugin-yubikey **resolves to executable in current directory** (.\age-plugin-yubikey.exe)
